connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultProcessInstanceContext.java

```diff
@@ -16,45 +16,69 @@
  */
 package io.camunda.connector.runtime.core.inbound;
 
-import io.camunda.connector.api.inbound.DynamicPropertyManager;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.camunda.connector.api.inbound.ProcessInstanceContext;
+import io.camunda.connector.api.validation.ValidationProvider;
+import io.camunda.connector.api.validation.ValidationUtil;
 import io.camunda.connector.feel.FeelEngineWrapper;
 import io.camunda.connector.feel.FeelEngineWrapperException;
 import java.util.HashMap;
 import java.util.Map;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-public class DefaultDynamicPropertyManager implements DynamicPropertyManager {
-  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultDynamicPropertyManager.class);
+public final class DefaultProcessInstanceContext implements ProcessInstanceContext {
+  private final Long processInstanceKey;
+  private final ValidationProvider validationProvider;
   private final FeelEngineWrapper feelEngineWrapper;
-  private final Map<Long, Map<String, Object>> propertiesById;
+  private final ObjectMapper objectMapper;
 
-  public DefaultDynamicPropertyManager(
+  private final Map<String, Object> properties;
+
+  public DefaultProcessInstanceContext(
+      final Long processInstanceKey,
+      final ValidationProvider validationProvider,
       final FeelEngineWrapper feelEngineWrapper,
-      final Map<Long, Map<String, Object>> propertiesById) {
+      final ObjectMapper objectMapper,
+      final Map<String, Object> properties) {
+    this.processInstanceKey = processInstanceKey;
+    if (validationProvider == null) {
+      this.validationProvider = ValidationUtil.discoverDefaultValidationProviderImplementation();
+    } else {
+      this.validationProvider = validationProvider;
+    }
     this.feelEngineWrapper = feelEngineWrapper;
-    this.propertiesById = propertiesById;
+    this.objectMapper = objectMapper;
+    this.properties = new HashMap<>(properties);
   }
 
-  public void setOrUpdateProperties(Long id, Map<String, Object> properties) {
-    propertiesById.putIfAbsent(id, new HashMap<>());
-    propertiesById.get(id).putAll(properties);
+  @Override
+  public void update(final Map<String, String> variables) {
+    properties.putAll(variables);
   }
```

Comment:
Why do you need this? You receive all properties via the intermediate context as its available right from the start when creating this instance?